homeassistant/components/media_source/media_source.py

```diff
@@ -0,0 +1,100 @@
+"""Local Media Source Implementation."""
+import datetime as dt
+import mimetypes
+from pathlib import Path
+
+from aiohttp import web
+
+from homeassistant.components.http import HomeAssistantView
+from homeassistant.components.http.static import CACHE_HEADERS
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN, MEDIA_MIME_TYPES, URI_SCHEME
+from .models import Media
+
+
+async def async_setup_media_source(hass: HomeAssistant):
+    """Set up local media source."""
+    hass.http.register_view(LocalMediaView(hass))
+    return ("local", async_find_media)
+
+
+async def async_find_media(hass: HomeAssistant, location=None) -> Media:
```

Comment:
Platforms need to get `media_id` passed in. Media Source should provide a parse_media_id function so that integrations can quickly get access to the info they need.

Media IDs could look like this:

```
media_source://integration/source_name/rest/of/path/bla.jpg
```

this could be parsed to

```python
class MediaId:
    integration: str = "media_source"
    source_name: str = "local"
    source_path: str = "rest/of/path/bla.jpg"
```

And then sources will just work with `source_path`. We could also just pass an instance of `MediaId` in. Media Source already parses it anyway to determine the integration.

Reply:
Now inside the Cast entity play media, it will be

```python
media_source_info = media_source.parse_media_id(media_id)

if media_source_info is not None:  # it's a media source media ID
    url = f"{hass_url}{media_source_info.get_api_path()}"
```

Reply:
(sorry, as I read your whole PR, I am slowly getting the full picture in my head)

Reply:
Ha okay I am getting myself confused. We should be able to already resolve this media_source_media_id to the url that we get from the source (in case of the local source the `/media/bla` url.) 

So we should not have a `media_source_info.get_api_path()` but maybe an 

```python
# Inside Cast Play Media
url = await media_source_info.async_get_url()

if url[0] == '/':  # relative path to HA
    url = f"{hass_url}{url}"
```